exe/Main.hs

```diff
@@ -190,22 +190,397 @@ showEvent lock (EventFileDiagnostics (toNormalizedFilePath' -> file) diags) =
+
+    let consultCradle :: Maybe FilePath -> FilePath -> IO (IdeResult HscEnvEq)
+        consultCradle hieYaml cfp = do
+           logInfo logger $ T.pack ("Consulting the cradle for " <> show cfp)
+           cradle <- maybe (loadImplicitCradle $ addTrailingPathSeparator dir) loadCradle hieYaml
+           eopts <- cradleToSessionOpts cradle cfp
+           logDebug logger $ T.pack ("Session loading result: " <> show eopts)
+           case eopts of
+             -- The cradle gave us some options so get to work turning them
+             -- into and HscEnv.
+             Right opts -> do
+               res <- session (hieYaml, toNormalizedFilePath' cfp, opts)
+               return res
+             -- Failure case, either a cradle error or the none cradle
+             Left err -> do
+               dep_info <- getDependencyInfo ([fp | Just fp <- [hieYaml]])
+               let ncfp = toNormalizedFilePath' cfp
+               let res = (map (renderCradleError ncfp) err, Nothing)
+               modifyVar_ fileToFlags $ \var -> do
+                 pure $ Map.insertWith HM.union hieYaml (HM.singleton ncfp (res, dep_info)) var
+               return res
+
+    -- This caches the mapping from hie.yaml + Mod.hs -> [String]
+    let sessionOpts :: (Maybe FilePath, FilePath) -> IO (IdeResult HscEnvEq)
+        sessionOpts (hieYaml, file) = do
+          v <- fromMaybe HM.empty . Map.lookup hieYaml <$> readVar fileToFlags
+          cfp <- liftIO $ canonicalizePath file
+          case HM.lookup (toNormalizedFilePath' cfp) v of
+            Just (opts, old_di) -> do
+              deps_ok <- checkDependencyInfo old_di
+              if not deps_ok
+                then do
+                  -- If the dependencies are out of date then clear both caches and start
+                  -- again.
+                  modifyVar_ fileToFlags (const (return Map.empty))
+                  -- Keep the same name cache
+                  modifyVar_ hscEnvs (return . Map.adjust (\(h, _) -> (h, [])) hieYaml )
+                  consultCradle hieYaml cfp
+                else return opts
+            Nothing -> consultCradle hieYaml cfp
+
+    dummyAs <- async $ return (error "Uninitialised")
+    runningCradle <- newVar dummyAs :: IO (Var (Async (IdeResult HscEnvEq)))
+    -- The main function which gets options for a file. We only want one of these running
+    -- at a time. Therefore the IORef contains the currently running cradle, if we try
+    -- to get some more options then we wait for the currently running action to finish
+    -- before attempting to do so.
+    let getOptions :: FilePath -> IO (IdeResult HscEnvEq)
+        getOptions file = do
+            hieYaml <- cradleLoc file
+            sessionOpts (hieYaml, file)
+    return $ \file -> do
+      liftIO $ join $ modifyVar runningCradle $ \as -> do
+        -- If the cradle is not finished, then wait for it to finish.
+        void $ wait as
+        as <- async $ getOptions file
```

Comment:
You can still get killed here afaict which will mean that the async will be running but you haven’t updated the var. I think you might want a `mask_` (but not uninterruptible) around the whole `modifyVar`? Then you can get interrupted during the `wait` but not between the `async` and updating the mvar.

Reply:
Yes, that seems to be what happens in `memoIO` as well. I am starting to look into fixing the tests and adding a test now if we are happy with the rest of the patch.